pytorch_lightning/core/datamodule.py

```diff
@@ -423,6 +423,14 @@ def __new__(cls, *args: Any, **kwargs: Any) -> "LightningDataModule":
         obj.prepare_data = cls._track_data_hook_calls(obj, obj.prepare_data)
         obj.setup = cls._track_data_hook_calls(obj, obj.setup)
         obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+
+        # mro defines the resolution order of superclasses during inheritence
+        # for a method call.
+        # We check that LightningDataModule is immediately next in line to
+        # the subclass 'cls', to make sure not to initialize any other subclass.
+        mro = cls.mro()
+        if mro[mro.index(cls) + 1] == LightningDataModule:
+            super(cls, obj).__init__()
```

Comment:
I don't understand why this `__new__` neds to call the `LightningDataModule.__init__`

Because that's what you are cheecking with that `if`, right?

Reply:
Yes. The `LightningDataModule.__init__` call initializes the private variables. If the user makes use of dataclass pattern and doesn't explicitly make the `super().__init__` call in the derived class, then we implicitly make the call in `__new__` without parameters. And in case the user does make the call somewhere (with or without the parameters) then it overrides the implicit call made in `__new__`.

`if mro[mro.index(cls) + 1] == LightningDataModule` is to make sure that incase the inheritence order is something like, `LightningDataModule->ABC->XYZ`, and `XYZ` is being initialized, we don't call `super().__init__` for it, which will point to `ABC`.

This makes sure that the users don't necessarily have to call init on LightningDataModule while using Dataclass pattern, but nothing breaks if they do.

Reply:
But if I delete this, the test added still passes.

If it's necessary, the test should check it.

Reply:
Added the test to check this.
Changing:
```
if mro[mro.index(cls) + 1] == LightningDataModule:
            super(cls, obj).__init__()
```
to:
```
super(cls, obj).__init__()
```
should make the `test_define_as_dataclass` fail, along with others.

Reply:
Tried again and if I remove these lines the datamodule tests still pass

Reply:
If you remove both the lines now, it would fail, because all the private variables won't be initialized.

If you just remove the `if` condition and run `super(cls, obj).__init__()`, it would fail because it wouldn't be checking for LightningDataModule as a direct parent.